### DISCUSSION TITLE
Can we make it more clear, as to how to get started with templates?

### DISCUSSION THREAD

**Discussion by jaredmeakin** [mockoon/mockoon#1149] [Help]
Hi Mockoon,

First off, thank you for making this amazing tool. It's been fun poking around, and figuring out how to mock endpoints.

However, one frustration I wanted to voice, as a new user, was that it's not immediately apparent how to create useful templates. Now, I 100% understand that predefined templates are you current business model. However, as a new user, the creation of templates was not easily approachable. I say this because there really were no good examples of "how" to create relevant templates.

Yes, you indicated `faker` was the underlying library, and later I found out about `mustache` through this discussion board--I some how missed it in the docs. However, it took me a trip to fakers docs, implementing current methods, and receiving errors... to realize I needed to be looking at the older docs. I spent quite a bit of time messing around with the mustache syntax. Including single-quotes, double-quotes, escaping single-quotes in double-quotes, and just being annoyed that I couldn't figure out how to properly format a date with faker (I later found out about the `date-fns` import). I also spent some time with `helpers.arrayElement()`, but couldn't get that to work. I ultimately landed on `"{{oneOf (array 'acitve' 'draft' 'archived')}}"` to do a simple `enum`.

These are definitely new comer problems, and something I'd be happy to help with. I think we could easily update existing documentation, with relevant examples, to make the usage of faker, mustache, and other helpers in Mockoon, something that the community as a whole can contribute to.

I could have easily implemented mock responses from an Express server via some factories. However, I really want to use this tool, and I see the value proposition that it holds.

All that being said, I'm one use case, and by no means do I intend to insult the team, their work, or this amazing product. I simply hope to provide a perspective, from fresh eyes, that might be helpful for adoption of this kick ass tool! I feel like a subscription model, for all users, isn't unfair for such a useful product. AI is cool, but I have a free ChatGPT 3.5 account. It works well enough if I feed it proper prompts. What's more important to me is building templates that mach ever changing business logic. That they can be developed quickly, and grouped by micro-service into logical folders. Also, I'll make a plug for OAuth and JWT authentication while I'm here. 😅

**Comment by 255kb** [comment]
Hi Jared,

It's true that the templating is a powerful feature but probably the hardest to discover. There are examples here and there, discussions on Discord or on GitHub and you may be able to find what you are looking for, but it may require some research even more when you don't know Handlebars' syntax. 
I'm all for improving the documentation with examples. If you are interested to contribute, you could add a new section in the documentation. It could compile examples found on the various discussions. I think it would be a good start. 
And if they are generic enough I would be glad to add them to the list of in-app templates too.

**Reply by jaredmeakin** [reply]
More than happy to compile the discussion board examples. It seems like you're the SME when it comes to templates. There are so many nuances to using handlebars, and I wish it was easier to understand them. Like which methods in faker are available. I was messing with `helpers.arrayElement()`, and couldn't get it to work. Or, I was trying `setVar` at the top of my template, but apparently if you name your variable the same thing as your property it doesn't work.

For example, this doesn't work. I messed with the syntax every which way following the provided documentation. Didn't work until I changed the name of my variable.... lol why?

```
{{ setVar 'status' (oneOf (array 'active' 'draft' 'archived')) }}
{{ setVar 'author' (faker 'datatype.uuid') }}
{
  posts:[
  {{#repeat 10}}
      {
      "id": "{{ faker 'datatype.uuid' }}",
      "status": "{{ @status }}",
      "content": "{{ faker 'lorem.paragraphs' 1}}",
      "created_at": "{{ date '2023-09-25' '2023-10-10' 'yyyy-MM-dd\'T\'HH:mm:ss.SSS\'Z\'' }}",
      "updated_at": "{{ date '2023-09-25' '2023-10-10' 'yyyy-MM-dd\'T\'HH:mm:ss.SSS\'Z\'' }}",
      "author_id": "{{ @author }}",
    }
    {{#unless @last}},{{/unless}}
  {{/repeat}}
  ]
}
```

**Reply by 255kb** [reply]
Faker is currently using v7. There is an opened [issue](https://github.com/mockoon/mockoon/issues/1150) to use v8. 

`arrayElement` should work with an array, something like: `{{ faker 'helpers.arrayElement' (array '1' '2' '3') }}`, but also if you fetch an array from a data bucket, request's body, query params, etc. with `bodyRaw 'path.to.my.array'`.

As for the `setVar` I see where this is coming from. [`status` is an existing helper](https://mockoon.com/docs/latest/templating/mockoon-response-helpers/). It's a bit unfortunate and maybe we should have namespaced the helpers. It's something we can think about for the future. Or namespace the variables to use them as `@var.myVarName`
